pkg/opsgenie/client.go

```diff
@@ -0,0 +1,105 @@
+package opsgenie
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"bitbucket.org/atlassianlabs/restclient"
+	"github.com/atlassian/voyager/pkg/util"
+	"github.com/atlassian/voyager/pkg/util/httputil"
+	"github.com/atlassian/voyager/pkg/util/pkiutil"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	asapAudience     = "micros-server"
+	asapSubject      = ""
+	integrationsPath = "/api/v1/opsgenie/integrations"
+)
+
+type Client struct {
+	logger     *zap.Logger
+	httpClient *http.Client
+	asap       pkiutil.ASAP
+	rm         *restclient.RequestMutator
+}
+
+func New(logger *zap.Logger, httpClient *http.Client, asap pkiutil.ASAP, baseURL *url.URL) *Client {
+	rm := restclient.NewRequestMutator(
+		restclient.BaseURL(baseURL.String()),
+	)
+	return &Client{
+		logger:     logger,
+		httpClient: httpClient,
+		asap:       asap,
+		rm:         rm,
+	}
+}
+
+// Gets OpsGenie integrations
+// return codes:
+// - 400: Bad request to Opsgenie
+// - 401: Unauthorized
+// - 404: Not found returned by Opsgenie. Does the specified Opsgenie team exist?
+// - 422: Semantic error in request to Opsgenie.
+// - 429: Rate limited by Opsgenie.
+func (c *Client) GetOrCreateIntegrations(ctx context.Context, teamName string) (*IntegrationsResponse, bool /* retriable */, error) {
+	req, err := c.rm.NewRequest(
+		pkiutil.AuthenticateWithASAP(c.asap, asapAudience, asapSubject),
+		restclient.Method(http.MethodGet),
+		restclient.JoinPath(fmt.Sprintf("%s/%s", integrationsPath, teamName)),
+		restclient.Context(ctx),
+		restclient.Header("Accept", "application/json"),
+	)
+	if err != nil {
+		return nil, false, errors.Wrap(err, "failed to create get integrations request")
+	}
+
+	response, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, false, errors.Wrap(err, "failed to execute get integrations request")
+	}
+	defer util.CloseSilently(response.Body)
+
+	retriable := false
```

Comment:
We really need a bit of code that determines where a response means a request can be retried.

The following could also be retriable:
```
http.StatusRequestTimeout
http.StatusServiceUnavailable
```

The error is also a `*url.Error` which has a `Temporary()` and `Timeout()` check to see what the cause might have been. I haven't been able to determine what causes a `Temporary()` or `Timeout()` to return true though.

I used them here in https://github.com/atlassian/voyager/pull/150/commits/732c738a80d6edc39281e468a96cb5cae9b4b7db#diff-3ec5620e3479cc95305dc75d504a817eR84 if you want to check it out.

Reply:
Yeah as discussed IRL I like the idea but would like to make those in a follow up PR unless we decide it blocks this PR.
My intention here was to address the main use cases where we want to retry.